handlers/smshandler/storage.go

```diff
@@ -0,0 +1,98 @@
+package smshandler
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/nyaruka/phonenumbers"
+	"github.com/vocdoni/blind-csp/types"
+)
+
+// ErrTooManyAttempts is returned when no more SMS attempts available for a user.
+var ErrTooManyAttempts = fmt.Errorf("too many SMS tries")
+
+// ErrUserAlreadyVerified is returned if the user is already verified when trying to verify it.
+var ErrUserAlreadyVerified = fmt.Errorf("user is already verified")
+
+// ErrUserNotBelongsToElection is returned if the user does not has participation rights for an election.
+var ErrUserNotBelongsToElection = fmt.Errorf("user does not belong to election")
+
+// ErrInvalidAuthToken is returned if the authtoken does not match with the election.
+var ErrInvalidAuthToken = fmt.Errorf("invalid authentication token")
+
+// ErrChallengeCodeFailure is returned when the challenge code does not match.
+var ErrChallengeCodeFailure = fmt.Errorf("challenge code do not match")
+
+// Users is the list of smshandler users.
+type Users struct {
+	Users map[string]UserData `json:"users"`
+}
+
+// UserData represents a user of the SMS handler.
+type UserData struct {
+	Elections []UserElection            `json:"elections,omitempty"`
+	ExtraData string                    `json:"extraData,omitempty"`
+	Phone     *phonenumbers.PhoneNumber `json:"phone,omitempty"`
+}
+
+type UserElection struct {
+	ElectionID        types.HexBytes `json:"electionId"`
+	RemainingAttempts int            `json:"remainingAttempts"`
+	Consumed          bool           `json:"consumed"`
+	AuthToken         *uuid.UUID     `json:"authToken,omitempty"`
+	Challenge         int            `json:"challenge,omitempty"`
+}
+
+// HexBytesToElection transforms a slice of HexBytes to []Election.
+// All entries are set with RemainingAttempts = attempts.
+func HexBytesToElection(electionIDs []types.HexBytes, attempts int) []UserElection {
+	elections := []UserElection{}
+
+	for _, e := range electionIDs {
+		ue := UserElection{}
+		ue.ElectionID = e
+		ue.RemainingAttempts = attempts
+		elections = append(elections, ue)
+	}
+	return elections
+}
+
+// FindElection returns the election index.
+// -1 is returned if the election ID is not found.
+func (ud *UserData) FindElection(electionID types.HexBytes) int {
+	for i, e := range ud.Elections {
+		if bytes.Equal(electionID, e.ElectionID) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Storage interface implements the storage layer for the smshandler
+type Storage interface {
```

Comment:
do we really need a storage abstraction when there's just one storage right now? I'm personally not a fan of doing it this way, because the interface rarely ends up being useful. For example, if we wanted to also support sqlite, we'd likely want a different interface to make some of the queries more efficient - this interface is aimed at key-value databases.

Reply:
There is a second implementation in mongodb (which requires an external server).

Reply:
Its ready, I'll push it ASAP.